HLTriggerOffline/Muon/src/HLTMuonValidator.cc

```diff
@@ -63,86 +51,92 @@ class HLTMuonValidator : public DQMEDAnalyzer {
   HLTMuonPlotter::L1MuonMatcherAlgoForDQM const l1tMuonMatcherAlgo_;
 };
 
-//////////////////////////////////////////////////////////////////////////////
-//////// Namespaces, Typedefs, and Constants /////////////////////////////////
+HLTMuonValidator::HLTMuonValidator(const edm::ParameterSet &pset)
+    : pset_(pset),
+      hltProcessName_(pset.getParameter<std::string>("hltProcessName")),
+      hltPathsToCheck_(pset.getParameter<std::vector<std::string>>("hltPathsToCheck")),
+      triggerEventToken_(consumes(edm::InputTag("hltTriggerSummaryRAW"))),
+      genParticlesToken_(consumes(pset.getParameter<std::string>("genParticleLabel"))),
+      recoMuonsToken_(consumes(pset.getParameter<std::string>("recMuonLabel"))),
+      l1tMuonMatcherAlgo_(pset, consumesCollector()) {}
 
-using namespace std;
-using namespace edm;
-using namespace reco;
-using namespace trigger;
+void HLTMuonValidator::fillLabels(std::string const &path,
+                                  std::vector<std::string> &moduleLabels,
+                                  std::vector<std::string> &stepLabels) const {
+  auto const &hltFilters = hltConfig_.saveTagsModules(path);
 
-typedef vector<string> vstring;
+  moduleLabels.clear();
+  moduleLabels.reserve(hltFilters.size());
 
-//////////////////////////////////////////////////////////////////////////////
-//////// Class Methods ///////////////////////////////////////////////////////
+  stepLabels.clear();
+  stepLabels.reserve(hltFilters.size() + 1);
 
-HLTMuonValidator::HLTMuonValidator(const ParameterSet &pset)
-    : pset_(pset),
-      hltProcessName_(pset.getParameter<string>("hltProcessName")),
-      hltPathsToCheck_(pset.getParameter<vstring>("hltPathsToCheck")),
-      triggerEventToken_(consumes(edm::InputTag("hltTriggerSummaryRAW"))),
-      genParticlesToken_(consumes(pset.getParameter<string>("genParticleLabel"))),
-      recoMuonsToken_(consumes(pset.getParameter<string>("recMuonLabel"))),
-      l1tMuonMatcherAlgo_(pset, consumesCollector()) {}
+  for (auto const &module : hltFilters) {
+    if (module.find("Filtered") == std::string::npos)
+      continue;
 
-vector<string> HLTMuonValidator::moduleLabels(string path) {
-  vector<string> modules = hltConfig_.moduleLabels(path);
```

Comment:
`moduleLabels` is not sufficient in this case, as there are modules called `*Filtered*` (in muon-related triggers of the HLT menu) which do not produce any `TriggerObjects` for `TriggerEventWithRefs` (see for example the module `hltIter0IterL3MuonPixelSeedsFromPixelTracksFiltered`); such extra modules would be taken into account by `HLTMuonPlotter`, fail the matching to GEN/RECO objects (because they return no HLT candidates), and thus cause the following HLT steps to fail the matching (even if one of the following HLT filters does have valid HLT candidates).

`saveTagsModules` (see above in green) should be a better choice (`saveTags` is a parameter present for all the modules in the HLT menu which are `HLTFilter`s, which do save the `TriggerObjects`).